### PR TITLE
validate metric and label names

### PR DESCRIPTION
**Pull request by ccmtaylor**
The regexes are copied from the Go client libraries; Go and the regex
crate both use RE2 syntax.

### CODE REVIEW

File: src/desc.rs [hunk 1]
```diff
@@ -77,15 +78,24 @@ impl Desc {
             return Err(Error::Msg("empty help string".into()));
         }
 
-        // TODO: check valid metric fq_name.
+        lazy_static! {
+            static ref VALID_METRIC_NAME: Regex = Regex::new(r"^[a-zA-Z_][a-zA-Z0-9_:]*$").unwrap();
```
Comment (siddontang) [inline src/desc.rs]: Is it better to move these two static vars outside Desc?

File: src/desc.rs [hunk 2]
```diff
@@ -77,15 +78,24 @@ impl Desc {
...
         label_values.push(fq_name);
 
         let mut label_names = BTreeSet::new();
 
         for label_name in const_labels.keys() {
-            // TODO: check invalid label name
+            if !VALID_LABEL_NAME.is_match(&label_name) {
+                return Err(Error::Msg(format!("'{}' is not a valid label name", &label_name)))
+            }
             if !label_names.insert(label_name.clone()) {
```
Comment (siddontang) [inline src/desc.rs]: can add a blank line.

File: src/desc.rs [hunk 3]
```diff
@@ -100,7 +110,9 @@ impl Desc {
         // cannot be in a regular label name. That prevents matching the label
         // dimension with a different mix between preset and variable labels.
         for label_name in &desc.variable_labels {
-            // TODO: check invalid label name
+            if !VALID_LABEL_NAME.is_match(&label_name) {
+                return Err(Error::Msg(format!("'{}' is not a valid label name", &label_name)))
+            }
             if !label_names.insert(format!("${}", label_name)) {
```
Comment (siddontang) [inline src/desc.rs]: ditto

**Comment by siddontang** [comment]
Thanks @ccmtaylor 

Rest LGTM
PTAL @overvenus

File: src/lib.rs [hunk 4]
```diff
@@ -22,6 +22,8 @@ extern crate fnv;
 extern crate lazy_static;
 extern crate hyper;
 extern crate libc;
+#[macro_use]
```
Comment (overvenus) [inline src/lib.rs]: Seems we don't need it's macros.

File: src/desc.rs [hunk 5]
```diff
@@ -137,3 +149,41 @@ impl Desc {
         Ok(desc)
     }
 }
+
+#[test]
```
Comment (overvenus) [inline src/desc.rs]: Put these tests in a `#[cfg(test)] mod tests` seems better.

File: src/desc.rs [hunk 6]
```diff
@@ -77,15 +78,24 @@ impl Desc {
...
 
         let mut label_values = Vec::with_capacity(const_labels.len() + 1);
         label_values.push(fq_name);
 
         let mut label_names = BTreeSet::new();
 
         for label_name in const_labels.keys() {
-            // TODO: check invalid label name
+            if !VALID_LABEL_NAME.is_match(&label_name) {
+                return Err(Error::Msg(format!("'{}' is not a valid label name", &label_name)))
```
Comment (overvenus) [inline src/desc.rs]: Add a `;` at the end of line 97, or `make format`

**Comment by overvenus** [comment]
Rest LGTM

File: src/desc.rs (hunk 6 above)
Comment (ccmtaylor) [inline src/desc.rs]: the Makefile doesn't have a `format` target. I've added semicolons and newlines manually.

Comment (ccmtaylor) [inline src/desc.rs]: erm, it does, sorry. Emacs didn't pick it up :blush:

**Pull request merged**